core/silkworm/consensus/blockchain.cpp

```diff
@@ -22,15 +22,19 @@
 
 namespace silkworm::consensus {
 
-Blockchain::Blockchain(State& state, ConsensusEngine& engine, const ChainConfig& config, const Block& genesis_block)
-    : state_{state}, engine_{engine}, config_{config} {
-    evmc::bytes32 hash{genesis_block.header.hash()};
-    state_.insert_block(genesis_block, hash);
-    state_.canonize_block(genesis_block.header.number, hash);
+Blockchain::Blockchain(State& state, const ChainConfig& config, const Block& genesis_block)
+    : state_{state}, config_{config}, engine_{consensus::engine_factory(config)} {
+    prime_state_with_genesis(genesis_block);
```

Comment:
Perhaps add
```
assert(engine_ != nullptr);
```

Reply:
Either a fully backed check for nullptr or leave as is. `asserts` completely loose their meaning on release builds (i.e. NDEBUG is defined)

Reply:
`assert` also serves as documentation. But yeah, at some point we can probably introduce a version of assert that is triggered in release builds as well. Let's leave as is for now.